File: sortingAlgo/bucketsort/BucketSort.cpp
// Bucket Sort in cpp

#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

void print_vec(vector<float> vec) {
   for(int i = 0; i<vec.size(); i++) {
      cout << vec[i] << " ";
   }
   cout << "\n";
}

void bucketSort(vector<float>&vec) {
   int size = vec.size();
   vector<float> bucket[size];

   // put elements into different buckets
   for(int i = 0; i<size; i++)  {          
      bucket[int(size * vec[i])].push_back(vec[i]);
   }

   //sort individual vectors
   for(int i = 0; i<size; i++) {
      sort(bucket[i].begin(), bucket[i].end());       
   }
   
   int index = 0;
   for(int i = 0; i < size; i++) {
      for (int j = 0; j < bucket[i].size(); j++) {
         vec[index++] = bucket[i][j]; 
      }
   }
}

int main() {
   
   vector<float> v(10);
   v = {0.542, 0.512, 0.761, 0.123, 0.541, .0981, 0.841, 0.01, 0.7, 0.901};
   
   cout << "Array before Sorting: ";
   print_vec(v);
   bucketSort(v);
   
   cout << "Array after Sorting: ";
   print_vec(v);
}

File: sortingAlgo/bucketsort/bucketSort.lua
--- Sort an array using the bucket sort algorithm.
--- @param t table
--- @return table
local function bucketSort(t)
    local n = #t
    local buckets = {}
    for i = 1, n do
        local bucket = math.floor(n * t[i])
        buckets[bucket] = buckets[bucket] or {}
        table.insert(buckets[bucket], t[i])
    end
    for i = 1, n do
        if buckets[i] then
            table.sort(buckets[i])
        end
    end
    local k = 1
    for i = 1, n do
        if buckets[i] then
            for j = 1, #buckets[i] do
                t[k] = buckets[i][j]
                k = k + 1
            end
        end
    end
    return t
end

return bucketSort